apps/server/src/shared/repo/base.do.repo.ts

```diff
@@ -10,45 +10,96 @@ export type EntityProperties<P> = P & IBaseEntityProps;
 export abstract class BaseDORepo<T extends BaseDO, E extends BaseEntity, P> {
 	constructor(protected readonly _em: EntityManager, protected readonly logger: Logger) {}
 
+	/**
+	 * Returns the name of the entity.
+	 */
 	abstract get entityName(): EntityName<E>;
 
+	/**
+	 * Returns a constructor for the entity.
+	 */
 	abstract getConstructor(): new (I) => E;
 
+	/**
+	 * Maps a domain object to its related entity properties.
+	 * @param entityDO The domain object
+	 * @returns The mapped entity properties
+	 * @protected
+	 */
 	protected abstract mapDOToEntity(entityDO: T): EntityProperties<P>;
+
+	/**
+	 * Maps a database entity to a domain object.
+	 * @param entity The database entity
+	 * @returns The domain object
+	 * @protected
+	 */
 	protected abstract mapEntityToDO(entity: E): T;
 
-	protected mapDOToEntityWithId(edo: T): EntityProperties<P> {
-		const entity: EntityProperties<P> = this.mapDOToEntity(edo);
-		entity.id = edo.id;
-		return entity;
+	/**
+	 * Maps a domain object to its related entity properties.
+	 * @param entityDO The domain object
+	 * @returns The mapped entity properties with id
+	 * @protected
+	 */
+	protected mapDOToEntityWithId(entityDO: T): EntityProperties<P> {
+		const entityProps: EntityProperties<P> = this.mapDOToEntity(entityDO);
+		entityProps.id = entityDO.id;
+		return entityProps;
 	}
 
+	/**
+	 * Uses the database entity constructor to create a new entity.
+	 * @param Type The constructor method for the entity
+	 * @param props The entity properties that are used to call the constructor
+	 * @returns A newly constructed database entity
+	 * @protected
+	 */
 	protected entityFactory(Type: new (I) => E, props: P): E {
 		return new Type(props);
 	}
 
-	async save(entityDos: T | T[]): Promise<T | T[]> {
-		const isArray = Array.isArray(entityDos);
-		const dos: T[] = isArray ? entityDos : [entityDos];
+	async save(entityDOs: T): Promise<T>;
```

Comment:
They are all needed? It look like parts are duplicated over the or condition?

async save(entityDOs: T): Promise<T>;
async save(entityDOs: T[]): Promise<T[]>;
async save(entityDOs: T | T[]): Promise<T | T[]>;
async save(entityDOs: T | T[]): Promise<T | T[]> { ... }

Reply:
I separated the save method into `save(T)` and `saveAll(T[])`, which gives enough type safety and gets rid of this.